### PR TITLE
perf(array): optimize array push back function performance

### PR DESCRIPTION
**Pull request by onecoolx**
### Description of the feature or fix

A clear and concise description of what the bug or new feature is.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Review by FASTSHIFT — DISMISSED** [review]

**Comment by onecoolx** [comment]
Decrease the number of memory allocations when array push back.
cc @XuNeo @FASTSHIFT

**Review by kisvegabor — DISMISSED** [review]

**Comment by kisvegabor** [comment]
I'm not sure if it's a good trade off. When a single event is added to a widget 4x8=32 bytes will be allocated instead of 4.

**Comment by onecoolx** [comment]
> Previously, we have discussed to add array runtime memory expanding options such as double the memory or linear expanding. But agreed in the end this is not necessary because array is not used in performance critical path.
> 
> #5379

This is not true. When you add points to a path in the vector api, resizing the memory each time will cause severe performance degradation and memory fragmentation. Strict memory growth is not the purpose of designing array container.

**Comment by onecoolx** [comment]
> I'm not sure if it's a good trade off. When a single event is added to a widget 4x8=32 bytes will be allocated instead of 4.

I think we can find a balance between memory growth and performance, that's why it's not growing by two times or more, but by a limited growth; this gives better performance than resizing every time you add an element, while the memory growth is manageable.

**Comment by kisvegabor** [comment]
I consider memory fragmentation a bigger issue than the performance. 


The 100 widgets need ~20kB.  

In case of the events, if we have these 100 widgets with 1 event, with the current method we will use 4 x 7 x 100 = 2.8kB extra RAM. (+14%) 
If we change `LV_ARRAY_DEFAULT_CAPACITY` to `4` it's only 1.2kB. (+6%) It seems like a good trade off to me.

**Comment by onecoolx** [comment]
> I consider memory fragmentation a bigger issue than the performance.
> 
> The 100 widgets need ~20kB.
> 
> In case of the events, if we have these 100 widgets with 1 event, with the current method we will use 4 x 7 x 100 = 2.8kB extra RAM. (+14%) If we change `LV_ARRAY_DEFAULT_CAPACITY` to `4` it's only 1.2kB. (+6%) It seems like a good trade off to me.

We could set this value become a build config , but as an array, when you add some values, you will most likely add more again.   For example  bezier curve need three points one time， when you create a complex path, memory will be realloc many times.

**Comment by kisvegabor** [comment]
Changing `LV_ARRAY_DEFAULT_CAPACITY` to `4`. I think it wouldn't be a good config option as it has too small impact. We should be able to find a good fixed value instead.

**Comment by FASTSHIFT** [comment]
We can set `LV_ARRAY_DEFAULT_CAPACITY` to 1 by default, which is more inclined to save memory. If the user has enough memory and wants to pursue speed, the user can increase the value of `LV_ARRAY_DEFAULT_CAPACITY` by himself

**Comment by kisvegabor** [comment]
I think 4 is a good default value a no need for bloating `lv_conf.h` even further. Although we can define it like
```
#ifndef LV_ARRAY_DEFAULT_CAPACITY
#define LV_ARRAY_DEFAULT_CAPACITY  4
#endif
```

So if someone really needs this level of fine tuning, it's still possible.

File: src/misc/lv_array.h [hunk 1]
```diff
@@ -18,7 +18,10 @@ extern "C" {
 /*********************
  *      DEFINES
  *********************/
-#define LV_ARRAY_DEFAULT_CAPACITY 8
+
+#ifndef LV_ARRAY_DEFAULT_CAPACITY
+#define LV_ARRAY_DEFAULT_CAPACITY  4
```
Comment (FASTSHIFT) [inline src/misc/lv_array.h:23]: Can we move this into `lv_conf`? That way it's configurable.

**Review by FASTSHIFT — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]

**Pull request merged**